### PR TITLE
Add estimagic enhancement proposal (EEP-01) for pytrees

### PR DESCRIPTION
**Pull request by janosg**

### CODE REVIEW

File: docs/source/development/eep-01-pytrees.rst [hunk 1]
```diff
@@ -0,0 +1,848 @@
...
+
+Needless to say, it is possible to register any iterable as container. For example, we
+would add pandas.Series and pandas.DataFrame (with varying definitions, depending on the
+application).
+
+
+Difference between pytrees in JAX and estimagic
+===============================================
+
+Most JAX functions `only work with Pytrees of arrays
```
Comment (hmgaudecker) [inline docs/source/development/eep-01-pytrees.rst:117]: Or arrays + scalars?

File: docs/source/development/eep-01-pytrees.rst [hunk 2]
```diff
@@ -0,0 +1,848 @@
...
+<https://jax.readthedocs.io/en/latest/pytrees.html#pytrees-and-jax-functions>`_, i.e.
+pytrees where container types are dicts, lists and tuples and all leaves are arrays.
+
+There are two ways to look at such pytrees:
```
Comment (hmgaudecker) [inline docs/source/development/eep-01-pytrees.rst:121]: The previous paragraph talked about Jax. I guess Jax takes the first route? 

*Aside: I would suggest avoiding the word "view" due to its typical meaning in the Python context, i.e., view vs. copy.*

File: docs/source/development/eep-01-pytrees.rst [hunk 3]
```diff
@@ -0,0 +1,848 @@
...
+2. As pytree of numbers -> ``tree_flatten`` produces a list of numbers
+
+Again, the difference between the two is which objects are registered as container types
+and the rules to flatten and unflatten them are defined.
+
+While numpy arrays, scalars and pandas.Series have only one natural way of defining
+the flattening rules, this becomes more complex for DataFrames due to the way ``params``
+DataFrames were used in estimagic before.
+
+We define the following rules: If a DataFrame contains a column called ``"value"``, we
```
Comment (hmgaudecker) [inline docs/source/development/eep-01-pytrees.rst:145]: My initial reaction was: What happens with bounds, then?

But really that thought should not arise. Thus make clearer somewhere what these guys are used for: For specifying start values / evaluations / etc., so the issue of flattening both dimensions of a DataFrame never arises, essentially only ever interested in a single Series of that DataFrame.

File: docs/source/development/eep-01-pytrees.rst [hunk 4]
```diff
@@ -0,0 +1,848 @@
...
+``"query"`` entry in the constraints dictionary.
+
+This keeps working as long as params are specified as a single DataFrame containing a
+``"value"`` column. If a more general pytree is used we need a "selector" entry instead.
+The value of that entry is a callable that takes the pytree and returns selected
+parameters.
+
+The selected parameters can be returned as estimagic-pytrees. For constraints where
+order plays a role (e.g., increasing), the order defined by ``tree_flatten`` is used.
+However we advise users to return one-dimensional arrays in that case, so they do not
```
Comment (hmgaudecker) [inline docs/source/development/eep-01-pytrees.rst:267]: The return of which function precisely is meant?

File: docs/source/development/eep-01-pytrees.rst [hunk 5]
```diff
@@ -0,0 +1,848 @@
...
+        criterion=crit,
+        params=params,
+        algorithm="scipy_lbfgsb",
+        constraints=constraints,
+    )
+
+Internally, constraints are already applied on a 1 dimensional numpy array and the
+parameter selections specified by "loc" and "query" are translated into positions in
+that array. What changes is that we now also have to translate the parameter selections
+from "selector" functions into positions. This is trivial by calling ``tree_unflatten``
```
Comment (hmgaudecker) [inline docs/source/development/eep-01-pytrees.rst:289]: Maybe too much detail on implementation, likely unclear for anyone but @janosg. Either drop or expand.

File: docs/source/development/eep-01-pytrees.rst [hunk 6]
```diff
@@ -0,0 +1,848 @@
...
+on an ``np.arange()`` of suitable length, calling the selector functions on the
+resulting pytree and recording all numbers that are there.
+
+
+Numerical derivatives during optimization
```
Comment (hmgaudecker) [inline docs/source/development/eep-01-pytrees.rst:294]: Combine with 'closed form' into one section. "derivatives during optimization" Just a short reference to the below? I find this first part rather unnecessary and the second on closed forms too short to be useful.

File: docs/source/development/eep-01-pytrees.rst [hunk 7]
```diff
@@ -0,0 +1,848 @@
...
+------------------------------------------------
+
+There are three types of moments in MSM estimation:
+
+- ``empirical moments``
+- The output of ``simulate_moments``
+- The output of ``calculate_moments``, needed to get a moments covariance matrix.
+
+We propose that moments can be stored as any valid estimagic pytree but of course all
+three types of moments have to be aligned, i.e. be stored in a tree of the same
```
Comment (hmgaudecker) [inline docs/source/development/eep-01-pytrees.rst:690]: Is there a utility for checking this? I saw something on equality comparisons, no time anymore to look it up right now, but I am not sure anymore whether that referred to equality of structures or equality of values in leafs.

**Review by hmgaudecker — COMMENTED** [review]
Really excellent work!!! I hope I did not mess anything up in f339c74, took the liberty to adjust wording while reading through it. I basically stopped just before the Derivatives section and only looked very briefly at estimation (looks great, though!!!).

File: docs/source/development/eep-01-pytrees.rst (hunk 3 above)
Comment (janosg) [reply docs/source/development/eep-01-pytrees.rst:145]: Made it clearer.

File: docs/source/development/eep-01-pytrees.rst (hunk 4 above)
Comment (janosg) [reply docs/source/development/eep-01-pytrees.rst:267]: Return of the selector function. It's clarified now.

File: docs/source/development/eep-01-pytrees.rst (hunk 5 above)
Comment (janosg) [reply docs/source/development/eep-01-pytrees.rst:289]: Replaced it by a shorter and more abstract description that mainly just states that the implementation will not be too hard.

File: docs/source/development/eep-01-pytrees.rst (hunk 6 above)
Comment (janosg) [reply docs/source/development/eep-01-pytrees.rst:294]: Done. Shortened it quite a bit. This is clear enough to guide the implementation but will require examples and tutorials for users to understand.

File: docs/source/development/eep-01-pytrees.rst (hunk 1 above)
Comment (janosg) [reply docs/source/development/eep-01-pytrees.rst:117]: Done.

File: docs/source/development/eep-01-pytrees.rst (hunk 2 above)
Comment (janosg) [reply docs/source/development/eep-01-pytrees.rst:121]: Replaced "view" by "perspective".

File: docs/source/development/eep-01-pytrees.rst (hunk 7 above)
Comment (janosg) [reply docs/source/development/eep-01-pytrees.rst:690]: Yes, can be checked. Added it.

**Review by hmgaudecker — APPROVED** [review]
Looks great to me! 

With the exception that to my mind, the milestones section does not really belong there. Maybe something for an issue, but I think of these foundational EEPs (or rather their equivalent in GETTSIM) to some extent as technical documentation, which will be there to stay.

**Review by timmens — APPROVED** [review]
I really like the idea of enhancement proposals and the way they democratize the decision process for big changes.

On top, I really like pytree's for estimagic. I think they will help estimagic stay up-to-date with the recent developments in state-of-the-art machine learning software like JAX.

**Pull request merged**